[4.0] New GLES3 Canvas Renderer

clayjohn opened the pull request:
This PR builds on the new shader compiler (https://github.com/godotengine/godot/pull/54946) and adds a batching renderer based on the (non-batching) Vulkan canvas renderer. 

Heavily WIP. Only supports rects and polygons right now. Performance is really good. Right now drawing the editor (with empty scene) takes about 0.5 ms on my device. For reference Vulkan is about 5-8 ms and 3.4 GLES3 is about 1.2.

Next steps are to:
1. finish off batching properly to support more features
2. Add the rest of the draw commands (multimesh, mesh, primitive, etc.)
3. then move onto advanced features (canvas group, backbuffer, particles etc.)

reduz committed:
WIP New GLES3 Shader Compiler

Uses versions and specializations (more similar to RenderingDevice version)

clayjohn committed:
New OpenGL batching canvas renderer

clayjohn commented:
Should be ready for a detailed review now. 

I have tested on my desktop and my laptop with intel integrated UHD 620 graphics. Performance on the integrated laptop is about the same as in 3.x.

lawnjelly commented:
I'll try and do a gradual better review (as it is large PR) but from when I started looking it looks very neat :+1: , much simplified over the canvas renderer in 3.x. I don't think it's necessary to have it "perfect" / all bugs sorted before merging first, otherwise we'd never get going, so will review on that basis.

A couple of little niggles I noticed so far in this screenshot:
![Screenshot from 2021-12-21 16-42-53](https://user-images.githubusercontent.com/21999379/146967365-0f8baf81-4d0b-4c67-9437-2a4781c2dbdf.png)
* I don't think the thick polylines is working correctly (probably index ordering differing from vulkan)
* The clipping to areas of the screen doesn't seem to work correctly in all cases, see the selection box for the node is drawn outside the main 2d window.

lawnjelly commented on drivers/gles3/rasterizer_gles3.h:
```diff
@@ -45,14 +44,13 @@ class RasterizerGLES3 : public RendererCompositor {
 	RasterizerCanvasGLES3 canvas;
 	RasterizerStorageGLES3 storage;
 	RasterizerSceneGLES3 scene;
 
```
I found if you swap the order to:
```
	RasterizerStorageGLES3 storage;
	RasterizerCanvasGLES3 canvas;
	RasterizerSceneGLES3 scene;
```
This cures the crash on exit, because you want the storage (and RID owners) to be constructed _before_ canvas and scene, and destroyed _after_ canvas and scene.

lawnjelly commented:
Have spent some time looking over this and it seems good to me, in terms of getting something merged and started. There are of course a few rough edges / todos but that is to be expected.

I've looked over mainly clayjohn's canvas renderer rather than reduz commit as I'm less familiar with that, but if they are being merged together I don't see why clayjohn shouldn't be able to review that.

There's a couple of small tweaks I mentioned above but apart from that it seems good to me (I don't know whether clayjohn was wanting to do more work on it first, but I'd be more in favour of getting something minimal merged). As usual the main caveat is just to make sure it won't affect the existing work of people who are using vulkan, but I'm presuming Calinou has mostly handled that side already.

clayjohn commented:
> I don't think the thick polylines is working correctly (probably index ordering differing from vulkan)

Fixed. Turns out I forgot to implement gl_primitives for polygons so our polylines were drawing using gl_triangles.

> The clipping to areas of the screen doesn't seem to work correctly in all cases, see the selection box for the node is drawn outside the main 2d window.

I haven't implemented clipping yet. I didn't plan on doing it in this PR, but I can.

> Also ERROR: Condition "!version" is true. at: version_set_code (drivers/gles3/shader_gles3.cpp:581) seems to be hitting quite a bit.

Fixed. Or rather, set code now exits early when called on a spatial shader. 

> Crash on exit in rid_owner:

Fixed. Thanks for the tip :)

clayjohn commented:
@lawnjelly. All fixed up and passing CI. 

@akien-mga I am fine with merging as-is if lawnjelly is as well. It is still missing many features. But it establishes a good base to build off of.

bruvzg reviewed (commented):
Tested on macOS, the editor is definitely much faster than with the old OpenGL renderer.

Few issues:
- 2D Sprites display random garbage (solid white/black or copy of the part of editor UI), with the following error:
```
ERROR: Condition "!tex->images.size()" is true. Returning: Ref<Image>()
   at: texture_2d_get (drivers/gles3/rasterizer_storage_gles3.cpp:601)
```

- Canvas selection/lines are drawn over other UI:
<img width="204" alt="Screenshot 2022-01-12 at 10 11 41" src="https://user-images.githubusercontent.com/7645683/149089720-7dc55471-e8bb-4554-aa6c-d7a21d8625f1.png">

- Tree lines seem to be misaligned:
<img width="162" alt="Screenshot 2022-01-12 at 10 12 10" src="https://user-images.githubusercontent.com/7645683/149089801-89799c23-ed37-4ca3-a99d-2385ebc34c87.png">

- Selecting any 3D Mesh or 2D mesh node crashes the editor.

lawnjelly commented:
I had a look at the crashing with 2d or 3d mesh above, and these are actually freezes in the generating previews it looks like, so not strictly speaking a part of the canvas renderer, just that these bits of code are not yet fault tolerant.

I think it's pretty expected it will go horribly wrong trying to create meshes at this stage, so am still happy to get this merged and get started with it.

akien-mga commented:
Alright, let's merge this as is. Please file bug reports for the issues you identified in the current implementation @bruvzg @lawnjelly.

akien-mga commented:
Thanks!
